<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="A tool for registering (aligning) gene expression profiles
    between two species (reference data and data to transform).">
<title>Gene Registration from Expression and Time-Courses in R â€¢ greatR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="deps/_Lato-0.4.1/font.css" rel="stylesheet">
<link href="deps/_Fira%20Code-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Gene Registration from Expression and Time-Courses in R">
<meta property="og:description" content="A tool for registering (aligning) gene expression profiles
    between two species (reference data and data to transform).">
<meta property="og:image" content="https://ruthkr.github.io/greatR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">greatR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Overview</h6>
    <a class="dropdown-item" href="articles/prepare-data.html">Preparing data</a>
    <a class="dropdown-item" href="articles/register-data.html">Registering data</a>
    <a class="dropdown-item" href="articles/visualise-results.html">Visualising results</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Optimisation</h6>
    <a class="dropdown-item" href="articles/optimise-parameters.html">Registering data with parameter optimisation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ruthkr/greatR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="greatr-">greatR <a class="anchor" aria-label="anchor" href="#greatr-"></a>
</h1>
</div>
<!-- badges: start -->
<p>The goal of <code>greatR</code> (<strong>G</strong>ene <strong>R</strong>egistration from <strong>E</strong>xpression <strong>a</strong>nd <strong>T</strong>ime-courses in <strong>R</strong>) to register (align) gene expression profiles between two species (reference data and data to transform). Non-reference gene expression profiles will be stretched and shifted. The optimality of registration parameters (shifts and stretches) will be estimated using least-squares criterion. This package is also designed to compare a registration model versus a non-registration model, as well as determine whether registration model performed better than non-registration transformation.</p>
<div class="section level2">
<h2 id="package-workflow">Package workflow<a class="anchor" aria-label="anchor" href="#package-workflow"></a>
</h2>
<p>Flowchart below illustrates the workflow of the package given an input data:</p>
<p><img src="reference/figures/greatR_workflow.png" width="85%"></p>
<p>More details on how to use this package are available on function documentations and vignettes:</p>
<ol style="list-style-type: decimal">
<li><a href="https://ruthkr.github.io/greatR/articles/prepare_data.html">Input requirements</a></li>
<li><a href="https://ruthkr.github.io/greatR/articles/register-data.html">Register data</a></li>
<li><a href="https://ruthkr.github.io/greatR/articles/visualise-results.html">Visualise registration results</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the released version of <code>greatR</code> from <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"greatR"</span><span class="op">)</span></code></pre></div>
<p>And the development version of <code>greatR</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ruthkr/greatR"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage---quick-start">Usage - quick start<a class="anchor" aria-label="anchor" href="#usage---quick-start"></a>
</h2>
<p>This is a basic example which shows you how to register (align) gene expression profiles over time:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ruthkr.github.io/greatR/">greatR</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load a data frame from the sample data</span>
<span class="va">all_data_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/brapa_arabidopsis_all_replicates.csv"</span>, package <span class="op">=</span> <span class="st">"greatR"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Running the registration</span>
<span class="va">registration_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scale_and_register_data.html">scale_and_register_data</a></span><span class="op">(</span>
  input_df <span class="op">=</span> <span class="va">all_data_df</span>,
  stretches <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2.5</span>, <span class="fl">2</span>, <span class="fl">1.5</span>, <span class="fl">1</span><span class="op">)</span>,
  shifts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">4</span>, <span class="fl">4</span>, length.out <span class="op">=</span> <span class="fl">33</span><span class="op">)</span>,
  min_num_overlapping_points <span class="op">=</span> <span class="fl">4</span>,
  initial_rescale <span class="op">=</span> <span class="cn">FALSE</span>,
  do_rescale <span class="op">=</span> <span class="cn">TRUE</span>,
  accession_data_to_transform <span class="op">=</span> <span class="st">"Col0"</span>,
  accession_data_ref <span class="op">=</span> <span class="st">"Ro18"</span>,
  start_timepoint <span class="op">=</span> <span class="st">"reference"</span>
<span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Information before registration â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; â„¹ Max value of expression_value of all_data_df: 262.28</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for all stretch and shift factor â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 3 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.6s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [85ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [91ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison BIC (10/10) [140ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 2.5 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.8s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [81ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [103ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison BIC (10/10) [160ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 2.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 2 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.9s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [95ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [82ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison BIC (10/10) [164ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 1.5 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [3s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [107ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [84ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison BIC (10/10) [171ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 1.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 1 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.7s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [85ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [90ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison BIC (10/10) [154ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Model comparison results â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; â„¹ BIC finds registration better than non-registration for: 10/10</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Applying the best-shifts and stretches to gene expression â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [85ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [98ms]</span>
<span class="co">#&gt; â„¹ Max value of expression_value: 9.05</span>
<span class="co">#&gt; âœ“ Imputing transformed expression values (10/10) [209ms]</span>
<span class="co">#&gt; </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>Calderwood, A., Hepworth, J., Woodhouse, â€¦ Morris, R. (2021). Comparative transcriptomics reveals desynchronisation of gene expression during the floral transition between Arabidopsis and Brassica rapa cultivars. <em>Quantitative Plant Biology, 2</em>, E4. <a href="https://www.cambridge.org/core/journals/quantitative-plant-biology/article/comparative-transcriptomics-reveals-desynchronisation-of-gene-expression-during-the-floral-transition-between-arabidopsis-and-brassica-rapa-cultivars/811BFDFA14F4BCC9C7F0ECC7CE103BB6" class="external-link">doi:10.1017/qpb.2021.6</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=greatR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/ruthkr/greatR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/ruthkr/greatR/issues/" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing greatR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1873-6203" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Alex Calderwood <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3749-035X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=greatR" class="external-link"><img src="https://www.r-pkg.org/badges/version/greatR" alt="CRAN_Status_Badge"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html" class="external-link"><img src="https://lifecycle.r-lib.org/articles/figures/lifecycle-stable.svg" alt="lifecycle"></a></li>
<li><a href="https://github.com/ruthkr/greatR/actions" class="external-link"><img src="https://github.com/ruthkr/greatR/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://app.codecov.io/gh/ruthkr/greatR" class="external-link"><img src="https://codecov.io/gh/ruthkr/greatR/branch/master/graph/badge.svg?token=L6TNLEPLLO" alt="codecov"></a></li>
<li><a href="https://github.com/ruthkr/greatR/actions/workflows/pkgdown.yaml" class="external-link"><img src="https://github.com/ruthkr/greatR/actions/workflows/pkgdown.yaml/badge.svg" alt="pkgdown"></a></li>
<li><a href="https://github.com/ruthkr/greatR/commits/master" class="external-link"><img src="https://img.shields.io/github/last-commit/ruthkr/greatR" alt="GitHub last commit"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a>, Alex Calderwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
