<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Gene Registration from Expression and Time-courses â€¢ GREAT</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Gene Registration from Expression and Time-courses">
<meta property="og:description" content="What the package does (one paragraph).">
<meta property="og:image" content="https://ruthkr.github.io/GREAT/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/prepare_data.html">Preparing data</a>
    </li>
    <li>
      <a href="articles/register-data.html">Registering data</a>
    </li>
    <li>
      <a href="articles/visualise-results.html">Visualising results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="great-">GREAT <img src="reference/figures/logo.png" align="right" width="120"><a class="anchor" aria-label="anchor" href="#great-"></a>
</h1></div>
<!-- badges: start -->

<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The goal of GREAT (Gene Registration from Expression and Time-courses) to register (align) gene expression profiles between two species (reference data and data to transform). Non-reference gene expression profiles will be stretched and shifted. The optimality of registration parameters (shifts and stretches) will be estimated using least-squares criterion. This package is also designed to compare a registration model versus a non-registration model, as well as determine whether registration model performed better than non-registration transformation.</p>
</div>
<div class="section level2">
<h2 id="package-workflow">Package workflow<a class="anchor" aria-label="anchor" href="#package-workflow"></a>
</h2>
<p>Flowchart below illustrates the workflow of the package given an input data:</p>
<p><img src="reference/figures/GREAT_workflow.png" width="85%"></p>
<p>More details on how to use this package are available on functions documentations and vignettes:</p>
<ol style="list-style-type: decimal">
<li><a href="https://ruthkr.github.io/GREAT/articles/prepare_data.html">Input requirements</a></li>
<li><a href="https://ruthkr.github.io/GREAT/articles/register-data.html">Register data</a></li>
<li><a href="https://ruthkr.github.io/GREAT/articles/visualise-results.html">Visualise registration results</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<!--
You can install the released version of GREAT from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("GREAT")
```
-->
<p>And the development version of <code>GREAT</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ruthkr/GREAT"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage---quick-start">Usage - quick start<a class="anchor" aria-label="anchor" href="#usage---quick-start"></a>
</h2>
<p>This is a basic example which shows you how to register (align) gene expression profiles over time:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GREAT</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Define the dataframe from the sample data</span>
<span class="co"># Gene expression data with replicates</span>
<span class="va">all_data_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/brapa_arabidopsis_all_replicates.csv"</span>, package <span class="op">=</span> <span class="st">"GREAT"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># Running the registration</span>
<span class="va">registration_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/scale_and_register_data.html">scale_and_register_data</a></span><span class="op">(</span>
    input_df <span class="op">=</span> <span class="va">all_data_df</span>,
    stretches <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">2.5</span>, <span class="fl">2</span>, <span class="fl">1.5</span>, <span class="fl">1</span><span class="op">)</span>,
    shift_extreme <span class="op">=</span> <span class="fl">4</span>,
    num_shifts <span class="op">=</span> <span class="fl">27</span>,
    min_num_overlapping_points <span class="op">=</span> <span class="fl">4</span>,
    initial_rescale <span class="op">=</span> <span class="cn">FALSE</span>,
    do_rescale <span class="op">=</span> <span class="cn">TRUE</span>,
    accession_data_to_transform <span class="op">=</span> <span class="st">"Col0"</span>,
    accession_data_ref <span class="op">=</span> <span class="st">"Ro18"</span>,
    start_timepoint <span class="op">=</span> <span class="st">"reference"</span>
  <span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Information before registration â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; â„¹ Max value of expression_value of all_data_df: 262.28</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for all stretch and shift factor â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 3 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [55ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [50ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison AIC &amp; BIC (10/10) [134ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 2.5 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.2s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [55ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [52ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison AIC &amp; BIC (10/10) [113ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 2.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 2 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2.2s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [47ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [50ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison AIC &amp; BIC (10/10) [99ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 1.5 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [2s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [51ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [52ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison AIC &amp; BIC (10/10) [102ms]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 1.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Analysing models for stretch factor = 1 â”€â”€</span>
<span class="co">#&gt; âœ“ Calculating score for all shifts (10/10) [1.9s]</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [49ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [49ms]</span>
<span class="co">#&gt; âœ“ Calculating registration vs non-registration comparison AIC &amp; BIC (10/10) [108ms]</span>
<span class="co">#&gt; âœ“ Finished analysing models for stretch factor = 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Model comparison results â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; â„¹ BIC finds registration better than non-registration for: 10/10</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; â”€â”€ Applying the best-shifts and stretches to gene expression â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">#&gt; âœ“ Normalising expression by mean and sd of compared values (10/10) [48ms]</span>
<span class="co">#&gt; âœ“ Applying best shift (10/10) [52ms]</span>
<span class="co">#&gt; â„¹ Max value of expression_value: 9.05</span>
<span class="co">#&gt; âœ“ Imputing transformed expression values (10/10) [144ms]</span>
<span class="co">#&gt; </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>Calderwood, A., Hepworth, J., Woodhouse, â€¦ Morris, R. (2021). Comparative transcriptomics reveals desynchronisation of gene expression during the floral transition between Arabidopsis and Brassica rapa cultivars. <em>Quantitative Plant Biology, 2</em>, E4. <a href="https://www.cambridge.org/core/journals/quantitative-plant-biology/article/comparative-transcriptomics-reveals-desynchronisation-of-gene-expression-during-the-floral-transition-between-arabidopsis-and-brassica-rapa-cultivars/811BFDFA14F4BCC9C7F0ECC7CE103BB6" class="external-link">doi:10.1017/qpb.2021.6</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GREAT</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1873-6203" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="lifecycle"></a></li>
<li><a href="https://codecov.io/gh/ruthkr/GREAT" class="external-link"><img src="https://codecov.io/gh/ruthkr/GREAT/branch/master/graph/badge.svg?token=L6TNLEPLLO" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/ruthkr/GREAT/actions/workflows/pkgdown.yaml" class="external-link"><img src="https://github.com/ruthkr/GREAT/actions/workflows/pkgdown.yaml/badge.svg" alt="pkgdown"></a></li>
<li><a href="https://github.com/ruthkr/GREAT/commits/master" class="external-link"><img src="https://img.shields.io/github/last-commit/ruthkr/GREAT" alt="GitHub last commit"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
