---
title: "Modifying load_mean_df functions"
output: 
  github_document:
    toc: true
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
library(genereg)
library(ggplot2)
library(data.table)
```

## Backgrounds

This Rmd is just for testing the new functions of new `load_mean_df` functions with the old one, just to make sure that I will not break the functions, or make sure it give the same results as the previous functions (using both `B.rapa` and `B. Oleracea`).

## Testing the functions to load data 

```{r}
b_oleracea_data <- readRDS("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/dh.rds")

# b_oleracea_data %>% 
#   head(2)
# 
# b_oleracea_data$sample_id <- b_oleracea_data$ballgown.dir
# b_oleracea_data$ballgown.dir <- NULL
# 
b_oleracea_data %>%
  head(2)
# 
# saveRDS(b_oleracea_data, "/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/dh_with_sample_id.rds")
```



```{r}
table_id_path <- "/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/ID_TAB.csv"
ara_path <- '/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/klepikova.rds'
b_oleracea_path <- "/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/dh_with_sample_id.rds"

# tools::file_ext("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/ID_TAB.csv")

brassica_test <- load_mean_df(file_path_brassica = b_oleracea_path, 
             file_path_arabidopsis = ara_path, 
             file_path_id_table = table_id_path, 
             tissue_wanted = "apex", 
             curr_GoIs = c("AT1G69120","AT5G61850","AT2G45660"), 
             sum_brassicas = F)


# brassica_get_all_data <- get_all_data(file_path_brassica = b_oleracea_path, 
#              file_path_arabidopsis = ara_path, 
#              file_path_id_table = table_id_path, 
#              col_names_wanted =  c("CDS.model", "sample_id", "FPKM", "accession", "tissue", "timepoint", "dataset", "est_counts", "group","norm.cpm"))
# 
# brassica_get_all_data %>% 
#   head(5)
```

```{r}
brassica_test[[1]]
```


## What to expect from the old functions 

```r
#old functions
get_data_all_symbol <- function(brassica_name) {

  rds_k <- 'klepikova' # 'klepikova

  # ------- changed by Ruth
  rdsdir <- '/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/'

  rdspath_k <- paste0(rdsdir, rds_k, '.rds')
  rdspath_b <- paste0(rdsdir, brassica_name, '.rds')

  expression_b <- readRDS(rdspath_b)
  expression_k <- readRDS(rdspath_k)

  # id data
  # ------- changed by Ruth
  ID_TABLE <- data.table::fread("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/ID_TAB.csv")

  ID_TABLE <- unique(ID_TABLE[, c('locus_name', 'symbol', 'CDS.model')])

  # this screws up the number of copies of each gene unless reorder based on symbol name first!
  #ID_TABLE <- ID_TABLE[order(ID_TABLE$symbol, decreasing=T), ]
  #ID_TABLE <- ID_TABLE[!duplicated(ID_TABLE[c('locus_name', 'CDS.model')]), ]

  # add ATG locus info and cut down
  little.IDT <- unique(ID_TABLE[, c('CDS.model', 'locus_name')]) %>% 
                dplyr::mutate(CDS.model = toupper(CDS.model)) %>% 
                dplyr::filter(!is.na(locus_name), !locus_name %in% c("", "-"))   # has to be unique, because ID_TABLE has duplicate rows for multiple symbols!!
  expression_b <- merge(expression_b, little.IDT, by='CDS.model')

  expression_k$locus_name <- expression_k$CDS.model

  # cut down to only have genes with ATG locus present in both datasets
  common_symbols <- intersect(expression_k$locus_name, expression_b$locus_name)
  expression_k <- expression_k[expression_k$locus_name %in% common_symbols, ]
  expression_b <- expression_b[expression_b$locus_name %in% common_symbols, ]
  
  # take a common columns
  colnames_intersect <- intersect(colnames(expression_k), colnames(expression_b))

  # join the two datasets into 1 & housekeeping
  expression <- rbind(expression_b[,..colnames_intersect], expression_k[,..colnames_intersect])

  # cut down to remove the 'blank' symbol
  expression <- expression[expression$locus_name!='',]

  return(expression)
}
```


## Test one by one 

```{r}
expression_b <- readRDS("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/dh.rds")
expression_k <- readRDS("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/klepikova.rds")

ID_TABLE <- data.table::fread("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/ID_TAB.csv")
ID_TABLE_all <- data.table::fread("/Users/kristiar/PhD/second-rotation-JIC/b_oleracea/data/ID_TAB_old.csv")

# add ATG locus info and cut down
little.IDT <- unique(ID_TABLE[, c('CDS.model', 'locus_name')]) %>% 
                dplyr::mutate(CDS.model = toupper(CDS.model)) %>% 
                dplyr::filter(!is.na(locus_name), !locus_name %in% c("", "-"))# has to be unique, because ID_TABLE has duplicate rows for multiple symbols!!
expression_b <- merge(expression_b, little.IDT, by='CDS.model')


expression_k$locus_name <- expression_k$CDS.model

expression_k %>% 
  head(5)

common_symbols <- intersect(expression_k$locus_name, expression_b$locus_name)

expression_k <- expression_k[expression_k$locus_name %in% common_symbols, ]
expression_b <- expression_b[expression_b$locus_name %in% common_symbols, ]

expression_k %>% 
  head(5)
```

```{r}
expression_b$locus_name %>% head()
```




```{r}
take_colnames_k <- expression_k %>% 
  colnames()

take_colnames_b <- expression_b %>% 
  colnames()

```


```{r}
test_in <- intersect(take_colnames_k, take_colnames_b)
```

```{r}
expression_k %>% 
  head(2)
```

```{r}
expression_b %>% 
  head(2)
```

```{r}
expression_b[, ..test_in] %>% 
  head(5)
```

```{r}

```


## Original old-function

```{r get_data_all_symbol}
get_data_all_symbol <- function(brassica_name) {

  rds_k <- 'klepikova' # 'klepikova

  # ------- changed by Ruth
  rdsdir <- 'final_data/rds/'

  rdspath_k <- paste0(rdsdir, rds_k, '.rds')
  rdspath_b <- paste0(rdsdir, brassica_name, '.rds')

  expression_b <- readRDS(rdspath_b)
  expression_k <- readRDS(rdspath_k)

  # id data
  # ------- changed by Ruth
  ID_TABLE <- readRDS(paste0('reference_data/ID_TABLE_brapa-v3.rds'))

  ID_TABLE <- unique(ID_TABLE[, c('locus_name', 'symbol', 'CDS.model')])

  # this screws up the number of copies of each gene unless reorder based on symbol name first!
  #ID_TABLE <- ID_TABLE[order(ID_TABLE$symbol, decreasing=T), ]
  #ID_TABLE <- ID_TABLE[!duplicated(ID_TABLE[c('locus_name', 'CDS.model')]), ]

  # add ATG locus info and cut down
  little.IDT <- unique(ID_TABLE[, c('CDS.model', 'locus_name')]) # has to be unique, because ID_TABLE has duplicate rows for multiple symbols!!
  expression_b <- merge(expression_b, little.IDT, by='CDS.model')

  #little.IDT <- ID_TABLE[, c('locus_name', 'symbol')]
  #names(little.IDT) <- c('CDS.model', 'symbol')
  #expression_k <- unique(merge(expression_k, little.IDT, by='CDS.model', allow.cartesian = T))
  expression_k$locus_name <- expression_k$CDS.model

  # cut down to only have genes with ATG locus present in both datasets
  common_symbols <- intersect(expression_k$locus_name, expression_b$locus_name)
  expression_k <- expression_k[expression_k$locus_name %in% common_symbols, ]
  expression_b <- expression_b[expression_b$locus_name %in% common_symbols, ]

  # join the two datasets into 1 & housekeeping
  expression <- rbind(expression_b, expression_k)

  # cut down to remove the 'blank' symbol
  expression <- expression[expression$locus_name!='',]

  # cut down to only have tissue present in both dataset
  #expression <- expression[expression$tissue=='apex', ]

  # cut down to only have similar/equivalent timepoint
  # cut down extra Col0 times
  #expression <- expression[expression$accession!='Col0' | expression$timepoint <=14, ]
  # cut down extra ZS11 times
  # expression <- expression[expression$accession!='ZS11' | (expression$timepoint<63 & expression$timepoint >=43), ]
  # # cut out dodgy 51 day timepoint
  # expression <- expression[expression$dataset!='ds_2018_06_19' | expression$timepoint != 51, ]

  rm(expression_b, expression_k, ID_TABLE, little.IDT)
  gc()

  return(expression)
}

test_rapa <- get_data_all_symbol('ro18_chiifu_apex')
```

```{r get_expression_oI}
get_expression_oI <- function(file_path_brassica, file_path_arabidopsis, file_path_id_table, curr_GoIs, sumBrassicas = FALSE) {

  # load rds and arabidopsis gene expression data into single df.
  master_exp <- get_all_data(file_path_brassica, file_path_arabidopsis, file_path_id_table)
  master_exp <- unique(master_exp)
  # cut down to common tissue
  exp <- master_exp[master_exp$tissue=='apex', ]
  #cut down to genes of interest (based on membership of comparison_genes.tsv)
  exp <- exp[exp$locus_name %in% curr_GoIs,]
  tmp <- unique(exp)
  tmp <- exp[exp$CDS.model=='BRAA06G036760.3C',]

  # different models can potentially have selected the same genes, but with diff symbol names - e.g. AP1, ATAP1 - want to ensure only have one of each
  # of these.
  # exp <- exp[order(exp$symbol, decreasing=F), ]
  # # get rid of different symbols for same genes really!
  # tmp <- exp[accession=='Col0', c('CDS.model', 'symbol')]
  # unique_tmp <- tmp[!duplicated(tmp[, c('CDS.model')])] # when multiple symbols for same CDS, keep one arbitrary one.
  # exp <- exp[exp$symbol %in% unique_tmp$symbol,]


  #length(unique(exp$symbol))
  #exp <- data.frame(exp)
  #exp <- exp[!duplicated(exp[c('sample_id', 'CDS.model')]), ]
  #length(unique(test$symbol))

  # reformat depending on how want to compare arabidopsis to brassica, using indiv. brassica genes, or summed brassica genes
  # if want to used summed brassica data to compare to the brassica: get symbol level expression total. Locus_name identity is ATG id
  if (sumBrassicas==T) {
    exp <- stats::aggregate(norm.cpm~sample_id+accession+tissue+timepoint+dataset+group+locus_name, data=exp, sum)
  } else if (sumBrassicas==F) {
    # otherwise duplicate each arabidopsis, so have an arabidopis copy for each brassica CDS gene. Now locus_name identity is CDS.model
    ara.exp <- exp[exp$accession=='Col0',]
    bra.exp <- exp[exp$accession!='Col0',]
    bra.ids <- unique(bra.exp[, c('CDS.model', 'locus_name')])
    ara.exp$CDS.model <- NULL
    ara.exp <- merge(bra.ids, ara.exp, by='locus_name', allow.cartesian=T)
    ara.exp$ara.id <- ara.exp$locus_name
    ara.exp$locus_name <- ara.exp$CDS.model
    bra.exp$ara.id <- bra.exp$locus_name
    bra.exp$locus_name <- bra.exp$CDS.model
    exp <- rbind(ara.exp, bra.exp)
  }

  # shorten experiment group names
  exp <- shorten_groups(exp)
  return(exp)
}
```


Now using new function, and compare with results with the old function

```{r test get_data_all_symbol}
rapa_path <- "final_data/rds/ro18_chiifu_apex.rds"
ara_path <- "final_data/rds/klepikova.rds"
table_rapa_path <- "reference_data/ID_TABLE_brapa-v3.rds"

test_new_func <- get_data_all_symbol(rapa_path, ara_path, table_rapa_path)

diff <- setdiff(test_rapa, test_new_func)

diff

class <- test_new_func %>% 
  class()

class[1]
class[2]
```

```{r get_all_data}
id_table_path <- "../inst/extdata/sample_data/id_table.RDS"
arabidopsis_expression_path <- "../inst/extdata/sample_data/arabidopsis_expression.RDS"
brassica_rapa_expression_path <- "../inst/extdata/sample_data/brassica_rapa_expression.RDS"

test_sample_data <- get_all_data(brassica_rapa_expression_path, arabidopsis_expression_path, id_table_path)

test_sample_data %>% 
  colnames()

```

```{r}
readRDS(brassica_rapa_expression_path) %>% 
  colnames()
```


```{r get_expression_of_interest}
id_table_path <- system.file("extdata/sample_data/id_table.RDS", package = "genereg")
arabidopsis_expression_path <- system.file("extdata/sample_data/arabidopsis_expression.RDS", package = "genereg")
brassica_rapa_expression_path <- system.file("extdata/sample_data/brassica_rapa_expression.RDS", package = "genereg")
list_gene_of_interest <- c("AT1G69120", "AT5G61850", "AT2G45660")

all_data <- get_expression_of_interest(brassica_rapa_expression_path,
                                       arabidopsis_expression_path,
                                       id_table_path,
                                       tissue_wanted = 'apex',
                                       curr_GoIs = list_gene_of_interest, 
                                       sum_brassicas = F)


all_data %>% 
  class()

all_data %>% 
  head(5)
```

```{r test get_expression_oI}
data_get_expression_oI <- get_expression_oI(brassica_rapa_expression_path, arabidopsis_expression_path, id_table_path, curr_GoIs = c("AT1G69120","AT5G61850","AT2G45660"),  sumBrassicas = F)


setdiff(data_get_expression_oI, all_data)
```
```{r load_mean.df}
load_mean.df <- function(file_path_brassica, file_path_arabidopsis, file_path_id_table, tissue_wanted, curr_GoIs, sum_brassicas = F) {
# 
#   #setwd('/Volumes/Research-Projects/bravo/alex/BRAVO_rna-seq/scripts/')
#   rds_file <- 'ro18_chiifu_apex' # don't include the .rds # the name of the brassica data to load
#   sumBrassicas <- F # if false use seperate brassica genes, and compare to repeated arabidopsis genes. If true, sume copies of each brassica and compare to arabidopsis
#   #datapath <- paste0('../final_data/rds/', rds_file, '.rds')
# 
# 
#   #### specify the genes to be used in the comparison: ---
#   # GoIs <- data.table::fread(paste0('../graphs/', rds_file, '/comparison_genes.tsv')) # read the list of Arabidopsis id's for the genes of interest
# 
#   # ------ changed by Ruth
#   GoIs <- data.table::fread(paste0('graphs/', rds_file, '/comparison_genes.tsv'))
# 
#   print(unique(GoIs$model))
#   keep_model_set <- c('rf') #, 'logistic', 'trough', 'spike', 'linear') # rf means don't force them to be the same AT ALL!
#   curr_GoIs <- GoIs$symbol[GoIs$model %in% keep_model_set]
#   # add some particular flowering genes of interest
#   # AT2G22540 : SVP
#   # AT3G57920 : SPL15
#   # AT2G42200 : SPL9
#   # AT5G51870 : AGL71 : BRAA10G010470.3C
#   # AT5G51860 : AGL72
#   # AT4G36920 : AP2
#   curr_GoIs <- c(curr_GoIs, 'AT2G22540', 'AT3G57920', 'AT2G42200', 'AT5G51870', 'AT5G51860', 'AT4G36920')
#   #curr_GoIs <- c('LMI1', 'STM', 'PNY', 'PNF', 'UFO', 'AGL42', 'AGL71', 'AGL72', 'PI', 'GA20OX1', 'GA20OX2') # KEY THINGS, MISSING



  ##### load the expression data for all the curr_GoIs gene models, for arabidopsis, and for the specified brassica
  # filt_models <- get_expression_oI(rds_file, curr_GoIs, sumBrassicas) #
  filt_models <- get_expression_of_interest(file_path_brassica, file_path_arabidopsis, file_path_id_table, tissue_wanted, curr_GoIs, sum_brassicas = F)
  length(unique(curr_GoIs))
  length(unique(filt_models$CDS.model))
  exp <- filt_models

  # get mean of each timepoint
  exp[, mean.cpm:=mean(norm.cpm), by=list(locus_name, accession, tissue, timepoint)]
  mean.df <- unique(exp[, c('locus_name', 'accession', 'tissue', 'timepoint', 'mean.cpm')])



  # filter mean.df to remove genes with very low expression - remove if max is less than 5, and less than half timepoints expressed
  # greater than 1
  ro18.df <- mean.df[mean.df$accession=='Ro18']
  ro18.df[, keep:=(max(mean.cpm) > 2 | mean(mean.cpm > 1) > 0.5) , by=.(locus_name)]
  keep.genes <- unique(ro18.df$locus_name[ro18.df$keep==TRUE])
  mean.df <- mean.df[mean.df$locus_name %in% keep.genes,]
  print(paste0(length(unique(mean.df$locus_name)), ' genes considered in the comparison'))
  rm(ro18.df, keep.genes)

  exp <- exp[exp$locus_name %in% unique(mean.df$locus_name)]
  # exp <- subset(exp, select=c('locus_name', 'accession', 'tissue', 'timepoint',
  #                             'norm.cpm', 'group'))
  # names(exp)[names(exp)=='norm.cpm'] <- 'mean.cpm'
  return(list(mean.df, exp))
}
```

```{r}
old_mean_data <- load_mean.df(file_path_brassica = brassica_rapa_expression_path, 
                              file_path_arabidopsis = arabidopsis_expression_path, 
                              file_path_id_table = id_table_path, 
                              tissue_wanted = "apex", 
                              curr_GoIs = list_gene_of_interest, 
                              sum_brassicas = F)
```

```{r}
old_mean_data[[1]]
```


```{r load_mean_df}
new_mean_data <- load_mean_df(file_path_brassica = brassica_rapa_expression_path, 
                              file_path_arabidopsis = arabidopsis_expression_path, 
                              file_path_id_table = id_table_path, 
                              tissue_wanted = "apex", 
                              curr_GoIs = list_gene_of_interest, 
                              sum_brassicas = F)
```


```{r}
setdiff(new_mean_data[[1]], old_mean_data[[1]])

setdiff(new_mean_data[[2]], old_mean_data[[2]])

new_mean_data[[1]] 


new_mean_data %>% class()
```























