---
title: "Preparing data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Preparing data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Input data frame required for `GREAT` to register your gene expression profiles is a data frame contains gene expression time-course data with all replicates. Illustrated diagram below shows the required `input_df` data frame format of `GREAT`. This data frame must contain:

- two different species which users wish to compare (for each gene, at least two replicates are required in all species), and
- six data columns: 

  - `locus_name`: accession or unique ID of each gene. 
  - `accession`: accession or name of the species to compare.
  - `timepoint`: time point of the gene expression data.
  - `expression_value`: desired RNA-seq expression value or measure of the abundance of gene or transcripts which one wishes to compare. This value can be RPM, RPKM, FPKM, TPM, TMM, DESeq, SCnorm, GeTMM, ComBat-Seq and raw reads counts.
  - `tissue`: tissue from which RNA-seq data were extracted. 
  - `group`: replicate ID for each expression value.
  
<br>

```{r all-data-illustration, echo=FALSE, out.width='100%'}
knitr::include_graphics("figures/all_data_illustration.jpg")
```

<br>
