---
title: "Test Alex's pipeline with his data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(data.table)
library(ggplot2)
library(viridis)
library(cowplot)
library(stringr)
library(splines)
library(dplyr)

source("../../dev/code_registration_from_paper/registration/registration_functions.R")
```

```{r}
##### SET PARAMETERS:

# parameters for performing registration:
do.initial.rescale <- 'false'  # 'rescale' to scale gene expression prior to registration.
do.register.rescale <- 'rescale'  # 'rescale' to scale gene expression using only overlapping timepoints
                                  # points during registration.
stretch = c(2, 1.5, 1) # candidate registration stretch factors to apply to Col-0
min.num.overlapping.points = 4 # will only consider shifts which leave at least this many overlapping timepoints 
                              # after applying the registration function.
shift.extreme  = 4 # the absolute maximum value which can be applied as a shift to gene expression timecourse (days).
transformed.timecourse = 'Col0' # the name of the timecourse to apply the registration function to.
num.shuffled = 1000  # the number of random permutations to use to test real results for significance.
shuffle.type <- 'shuffle.genes' # method for random permutation of the data to assess significance of real results. 
                                # 'shuffle.genes' : randomly reallocate orthologues in Arabdidopsis compared to 
                                # each R-o-18 gene.
in.data.path <- './data_for_registration.rds'  # path to gene expression data
out.dir <- './registration_out/'  # path to directory to save results

##### END PARAMETERS


# create out.dir if doesn't exist
dir.create(out.dir, showWarnings = F)

# convert flags for rescaling to bools.
if (do.initial.rescale=='rescale') {
  initial.rescale <- TRUE
} else {
  initial.rescale <- FALSE
}
if (do.register.rescale=='rescale') {
  should.rescale <- TRUE
} else {
  should.rescale <- FALSE
}


##### LOAD GENE EXPRESSION DATA:
# load expression data for compared genes.

# mean.df : the mean gene expression of each gene in each genotype at each timepoint. 
# Columns:
# "locus_name" : brassica gene identifier. (identify of brassica genes. identity of brassica orthologue of Arabidopsis genes)
# "accession" : genotype - "Col0" or "Ro18"
# "timepoint" : the days post germination the sample taken
# "mean.cpm" : mean of TMM normalised counts per million amoung biological replicates.

# all.data.df : all replicates of gene expression in each genotype at each timepoint.
# Columns: 
# "locus_name" : brassica gene identifier. (identify of brassica genes. identity of brassica orthologue of Arabidopsis genes)
# "accession" : genotype - "Col0" or "Ro18"
# "timepoint" : the days post germination the sample taken
# "group" : library identifier, format "genotype-timepoint-replicate"
# "mean.cpm" : TMM normalised counts per million in each biological replicate (group).


mean.df <- filtered_alex_data_mean

all.data.df <- filtered_alex_data_all


O <- prepare_scaled_and_registered_data(mean.df, all.data.df,
  stretch = stretch, initial.rescale, should.rescale, min.num.overlapping.points,
  shift.extreme, transformed.timecourse
)
```

```{r}
id_table_path <- system.file("extdata/sample_data/id_table_5genes.RDS", package = "GREAT")


ID_table <- readRDS(id_table_path)
ID_table %>% 
  dplyr::select(CDS.model, position, locus_name, gene_id, symbol) %>% 
  # dplyr::arrange(symbol)
  # dplyr::filter(locus_name %in% c("AT1G69120", "AT5G61850", "AT2G45660")) %>% 
  dplyr::pull(locus_name) %>% 
  unique()

model_compared <- O[['model.comparison']] %>% 
  dplyr::left_join(ID_table %>% 
          dplyr::mutate(CDS.model = toupper(CDS.model)), by = c("gene" = "CDS.model")) %>% 
  dplyr::select(c(gene, symbol, stretch, shift))
  

test_mapped <- O[['imputed.mean.df']] %>%
  dplyr::left_join(ID_table %>% 
                     dplyr::mutate(CDS.model = toupper(CDS.model)), by = c("locus_name" = "CDS.model")) %>% 
  dplyr::select(-c(locus_name.y)) %>%
  dplyr::rename(locus_name = symbol, bra_gene = locus_name)


test <- test_mapped %>% 
  dplyr::filter(dplyr::case_when(locus_name == "AGL24" ~ shifted.time >= 10 & shifted.time <= 16, 
                                 locus_name == "AP1" ~  shifted.time >= 11 & shifted.time <= 30,
                                 locus_name == "AP3" ~  shifted.time >= 13 & shifted.time <= 25,
                                 locus_name == "LFY" ~ shifted.time >= 11 & shifted.time <= 25,
                                 locus_name == "SOC1" ~ shifted.time >= 10 & shifted.time <= 30))

test <- test_mapped %>% 
  dplyr::filter(dplyr::case_when(locus_name == "AGL24" ~ shifted.time >= 14 & shifted.time <= 35, 
                          locus_name == "AP1" ~  shifted.time >= 11 & shifted.time <= 25,
                          locus_name == "AP3" ~  shifted.time >= 13 & shifted.time <= 25,
                          locus_name == "LFY" ~ shifted.time >= 11 & shifted.time <= 19,
                          locus_name == "SOC1" ~ shifted.time >= 10 & shifted.time <= 31))

test

plot_registered_GoIs_for_comparible_timepoints(test_mapped)

plot_registered_GoIs_for_comparible_timepoints(test)
```



