<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="greatR">
<title>Preparing data • greatR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Lato-0.4.1/font.css" rel="stylesheet">
<link href="../deps/_Fira%20Code-0.4.1/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Preparing data">
<meta property="og:description" content="greatR">
<meta property="og:image" content="https://ruthkr.github.io/greatR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">greatR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Overview</h6>
    <a class="dropdown-item" href="../articles/prepare-data.html">Preparing data</a>
    <a class="dropdown-item" href="../articles/register-data.html">Registering data</a>
    <a class="dropdown-item" href="../articles/visualise-results.html">Visualising results</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Optimisation</h6>
    <a class="dropdown-item" href="../articles/optimise-parameters.html">Registering data with parameter optimisation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ruthkr/greatR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="prepare-data_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Preparing data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ruthkr/greatR/blob/HEAD/vignettes/prepare-data.Rmd" class="external-link"><code>vignettes/prepare-data.Rmd</code></a></small>
      <div class="d-none name"><code>prepare-data.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="data-requirements">Data requirements<a class="anchor" aria-label="anchor" href="#data-requirements"></a>
</h2>
<p>Input data frame required for <code>greatR</code> to register your gene expression profiles is a data frame that contains gene expression time-course data with all replicates. The illustrated diagram below shows the required <code>input_df</code> data frame format of <code>greatR</code>. This data frame must contain:</p>
<ul>
<li><p>two different species which users wish to compare (for each gene, at least two replicates are required in all species), and</p></li>
<li>
<p>six data columns:</p>
<ul>
<li>
<code>locus_name</code>: accession or unique ID of each gene.</li>
<li>
<code>accession</code>: accession or name of the species to compare.</li>
<li>
<code>timepoint</code>: time point of the gene expression data.</li>
<li>
<code>expression_value</code>: desired RNA-seq expression value or measure of the abundance of gene or transcripts which one wishes to compare. This value can be RPM, RPKM, FPKM, TPM, TMM, DESeq, SCnorm, GeTMM, ComBat-Seq and raw reads counts.</li>
<li>
<code>tissue</code>: tissue from which RNA-seq data were extracted.</li>
<li>
<code>group</code>: replicate ID for each expression value.</li>
</ul>
</li>
</ul>
<p><img src="figures/all_data_illustration.jpg" width="80%" style="display: block; margin: auto;"></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="using-a-helper-function-to-prepare-input-data">Using a helper function to prepare input data<a class="anchor" aria-label="anchor" href="#using-a-helper-function-to-prepare-input-data"></a>
</h2>
<p><code>greatR</code> package provides a helper function <code><a href="../reference/get_expression_of_interest.html">get_expression_of_interest()</a></code> to combine reference and query (data to register) data frames, if table ID which connect reference and query data frames provided (as illustrated in the figure below).</p>
<p><img src="figures/prepare_data.png" width="100%" style="display: block; margin: auto;"></p>
<p><br></p>
<p>This package also provides data frames as examples to use this function. Those data frames are expression data of <em>B. rapa</em> (<a href="https://www.cambridge.org/core/journals/quantitative-plant-biology/article/comparative-transcriptomics-reveals-desynchronisation-of-gene-expression-during-the-floral-transition-between-arabidopsis-and-brassica-rapa-cultivars/811BFDFA14F4BCC9C7F0ECC7CE103BB6" class="external-link">Calderwood et al., 2021</a>), <em>Arabidopsis</em> (<a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-015-1688-9" class="external-link">Klepikova et al., 2015</a>), and join table or table ID which contains both accession of <em>B. rapa</em> and <em>Arabidopsis</em>. This table ID is important to combine both <em>B. rapa</em> and <em>Arabidopsis</em> data frames.</p>
<div class="section level3">
<h3 id="load-and-view-example-data">Load and view example data<a class="anchor" aria-label="anchor" href="#load-and-view-example-data"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ruthkr.github.io/greatR/">greatR</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></code></pre></div>
<p>To load example data frames provided within the package, users need to run the following procedure:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load example data</span>
<span class="va">id_table_rapa_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata/sample_data/id_table_5genes.rds"</span>, 
  package <span class="op">=</span> <span class="st">"greatR"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">klepikova_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata/sample_data/arabidopsis_expression_5genes.rds"</span>, 
  package <span class="op">=</span> <span class="st">"greatR"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">rapa_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>
  <span class="st">"extdata/sample_data/brassica_rapa_expression_5genes.rds"</span>, 
  package <span class="op">=</span> <span class="st">"greatR"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Let’s take a look on each data frame we have loaded below:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># View reference data</span>
<span class="va">rapa_subset</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="13%">
<col width="21%">
<col width="7%">
<col width="7%">
<col width="5%">
<col width="7%">
<col width="10%">
<col width="8%">
<col width="10%">
<col width="7%">
</colgroup>
<thead><tr class="header">
<th align="left">CDS.model</th>
<th align="left">sample_id</th>
<th align="right">FPKM</th>
<th align="left">accession</th>
<th align="left">tissue</th>
<th align="right">timepoint</th>
<th align="left">dataset</th>
<th align="right">est_counts</th>
<th align="left">group</th>
<th align="right">norm.cpm</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_1</td>
<td align="right">0.287365</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">12</td>
<td align="left">Ro18-apex-11</td>
<td align="right">0.3968734</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_2</td>
<td align="right">1.069565</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">47</td>
<td align="left">Ro18-apex-11</td>
<td align="right">1.4147711</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_3</td>
<td align="right">0.541893</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">23</td>
<td align="left">Ro18-apex-11</td>
<td align="right">0.7423984</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># View query data</span>
<span class="va">klepikova_subset</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="8%">
<col width="22%">
<col width="7%">
<col width="8%">
<col width="5%">
<col width="8%">
<col width="12%">
<col width="8%">
<col width="10%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">CDS.model</th>
<th align="left">sample_id</th>
<th align="right">FPKM</th>
<th align="left">accession</th>
<th align="left">tissue</th>
<th align="right">timepoint</th>
<th align="left">dataset</th>
<th align="right">est_counts</th>
<th align="left">group</th>
<th align="right">norm.cpm</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AT1G69120</td>
<td align="left">ERR_ds_klepikova_SRR1659921</td>
<td align="right">0.299101</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">7</td>
<td align="left">ds_klepikova_1</td>
<td align="right">4</td>
<td align="left">Col0-apex-07</td>
<td align="right">0.4667855</td>
</tr>
<tr class="even">
<td align="left">AT1G69120</td>
<td align="left">ERR_ds_klepikova_SRR1659922</td>
<td align="right">0.020064</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">7</td>
<td align="left">ds_klepikova_2</td>
<td align="right">1</td>
<td align="left">Col0-apex-07</td>
<td align="right">0.0741901</td>
</tr>
<tr class="odd">
<td align="left">AT1G69120</td>
<td align="left">ERR_ds_klepikova_SRR1660397</td>
<td align="right">0.000000</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">8</td>
<td align="left">ds_klepikova_1</td>
<td align="right">0</td>
<td align="left">Col0-apex-08</td>
<td align="right">0.0000000</td>
</tr>
</tbody>
</table>
<p>As we can see, both data frames of expression profiles above contain a column of gene accession (<code>CDS.model</code>). These two columns from each data frame were mapped into a table ID as we can see below: <code>CDS.model</code> for <em>B. rapa</em> gene accessions and <code>locus_name</code> for <em>Arabidopsis</em> gene accessions.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># View table ID</span>
<span class="va">id_table_rapa_subset</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">CDS.model</span>, <span class="va">locus_name</span>, <span class="va">symbol</span>, <span class="va">gene_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">CDS.model</th>
<th align="left">locus_name</th>
<th align="left">symbol</th>
<th align="left">gene_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA01G016140.3C</td>
<td align="left">AT4G24540</td>
<td align="left">AGL24</td>
<td align="left">AT4G24540</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">AT1G69120</td>
<td align="left">AP1</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G043220.3C</td>
<td align="left">AT5G61850</td>
<td align="left">LFY</td>
<td align="left">AT5G61850</td>
</tr>
</tbody>
</table>
<p>Having all of these informations, we can then easily get the required input data.</p>
</div>
<div class="section level3">
<h3 id="combine-data">Combine data<a class="anchor" aria-label="anchor" href="#combine-data"></a>
</h3>
<p>Users can use function <code><a href="../reference/get_expression_of_interest.html">get_expression_of_interest()</a></code> to combine two different table data frames containing expression values that users wish to compare.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_expression_of_interest.html">get_expression_of_interest</a></span><span class="op">(</span>
  data_ref <span class="op">=</span> <span class="va">rapa_subset</span>,
  data_to_transform <span class="op">=</span> <span class="va">klepikova_subset</span>,
  id_table <span class="op">=</span> <span class="va">id_table_rapa_subset</span>,
  lookup_col_ref_and_id_table <span class="op">=</span> <span class="st">"CDS.model"</span>,
  lookup_col_ref_and_to_transform <span class="op">=</span> <span class="st">"locus_name"</span>,
  colnames_wanted <span class="op">=</span> <span class="cn">NULL</span>,
  tissue_wanted <span class="op">=</span> <span class="st">"apex"</span>,
  gene_of_interest_acc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AT1G69120"</span>, <span class="st">"AT5G618"</span><span class="op">)</span>,
  sum_exp_data_ref <span class="op">=</span> <span class="cn">FALSE</span>,
  accession_data_to_transform <span class="op">=</span> <span class="st">"Col0"</span>
<span class="op">)</span></code></pre></div>
<p>Let’s take a look on how the combined data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># View data</span>
<span class="va">all_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">accession</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="10%">
<col width="10%">
<col width="16%">
<col width="5%">
<col width="6%">
<col width="4%">
<col width="6%">
<col width="9%">
<col width="6%">
<col width="7%">
<col width="6%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th align="left">locus_name</th>
<th align="left">CDS.model</th>
<th align="left">sample_id</th>
<th align="right">FPKM</th>
<th align="left">accession</th>
<th align="left">tissue</th>
<th align="right">timepoint</th>
<th align="left">dataset</th>
<th align="right">est_counts</th>
<th align="left">group</th>
<th align="right">norm.cpm</th>
<th align="left">id_transform_data</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ds_klepikova_SRR1659921</td>
<td align="right">0.299101</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">7</td>
<td align="left">ds_klepikova_1</td>
<td align="right">4</td>
<td align="left">Col0-apex-07</td>
<td align="right">0.4667855</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ds_klepikova_SRR1659922</td>
<td align="right">0.020064</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">7</td>
<td align="left">ds_klepikova_2</td>
<td align="right">1</td>
<td align="left">Col0-apex-07</td>
<td align="right">0.0741901</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ds_klepikova_SRR1660397</td>
<td align="right">0.000000</td>
<td align="left">Col0</td>
<td align="left">apex</td>
<td align="right">8</td>
<td align="left">ds_klepikova_1</td>
<td align="right">0</td>
<td align="left">Col0-apex-08</td>
<td align="right">0.0000000</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_1</td>
<td align="right">0.287365</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">12</td>
<td align="left">Ro18-apex-11</td>
<td align="right">0.3968734</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_2</td>
<td align="right">1.069565</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">47</td>
<td align="left">Ro18-apex-11</td>
<td align="right">1.4147711</td>
<td align="left">AT1G69120</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">BRAA02G018970.3C</td>
<td align="left">ERR_ro18_rna_seq_v3_R18A1_3</td>
<td align="right">0.541893</td>
<td align="left">Ro18</td>
<td align="left">apex</td>
<td align="right">11</td>
<td align="left">ro18_rna_seq</td>
<td align="right">23</td>
<td align="left">Ro18-apex-11</td>
<td align="right">0.7423984</td>
<td align="left">AT1G69120</td>
</tr>
</tbody>
</table>
<p>After combining reference data and data to transform, users need to adjust it to the required input data, this can be done as follows.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">all_data_final</span> <span class="op">&lt;-</span> <span class="va">all_data</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>
    <span class="va">locus_name</span>, 
    <span class="va">accession</span>, 
    <span class="va">timepoint</span>, 
    expression_value <span class="op">=</span> <span class="va">norm.cpm</span>, 
    <span class="va">tissue</span>, 
    <span class="va">group</span>
  <span class="op">)</span>

<span class="va">all_data_final</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">accession</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span>
  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">locus_name</th>
<th align="left">accession</th>
<th align="right">timepoint</th>
<th align="right">expression_value</th>
<th align="left">tissue</th>
<th align="left">group</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">7</td>
<td align="right">0.4667855</td>
<td align="left">apex</td>
<td align="left">Col0-apex-07</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">7</td>
<td align="right">0.0741901</td>
<td align="left">apex</td>
<td align="left">Col0-apex-07</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">8</td>
<td align="right">0.0000000</td>
<td align="left">apex</td>
<td align="left">Col0-apex-08</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">8</td>
<td align="right">0.0000000</td>
<td align="left">apex</td>
<td align="left">Col0-apex-08</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">9</td>
<td align="right">0.3722542</td>
<td align="left">apex</td>
<td align="left">Col0-apex-09</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">9</td>
<td align="right">0.0000000</td>
<td align="left">apex</td>
<td align="left">Col0-apex-09</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">0.3968734</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-11</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">1.4147711</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-11</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">0.7423984</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-11</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">11.3007002</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-29</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">23.2055664</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-29</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">22.0307747</td>
<td align="left">apex</td>
<td align="left">Ro18-apex-29</td>
</tr>
</tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a>, Alex Calderwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
