<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="greatR">
<title>Registering data • greatR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Lato-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Fira_Code-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Registering data">
<meta property="og:description" content="greatR">
<meta property="og:image" content="https://ruthkr.github.io/greatR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">greatR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Overview</h6>
    <a class="dropdown-item" href="../articles/data-requirement.html">Input data requirements</a>
    <a class="dropdown-item" href="../articles/register-data.html">Registering data</a>
    <a class="dropdown-item" href="../articles/process-results.html">Processing results</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Others</h6>
    <a class="dropdown-item" href="../articles/register-data-manually.html">Registering data without optimisation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ruthkr/greatR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Registering data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ruthkr/greatR/blob/HEAD/vignettes/register-data.Rmd" class="external-link"><code>vignettes/register-data.Rmd</code></a></small>
      <div class="d-none name"><code>register-data.Rmd</code></div>
    </div>

    
    
<p>This article will show users how to register data using different
types of input, as illustrated below.</p>
<p><img src="figures/registration_opt_diagram.png" width="100%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="using-data-frame-input">Using data frame input<a class="anchor" aria-label="anchor" href="#using-data-frame-input"></a>
</h2>
<div class="section level3">
<h3 id="loading-sample-data">Loading sample data<a class="anchor" aria-label="anchor" href="#loading-sample-data"></a>
</h3>
<p><code>greatR</code> provides an example of data frame containing two
different species <em>A. thaliana</em> and <em>B. rapa</em> with two and
three different replicates, respectively. This data frame can be read as
follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load the package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ruthkr.github.io/greatR/">greatR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load a data frame from the sample data</span></span>
<span><span class="va">b_rapa_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/brapa_arabidopsis_data.csv"</span>, package <span class="op">=</span> <span class="st">"greatR"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Note that the data has all of five columns required by the
package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">b_rapa_data</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">gene_id</th>
<th align="left">accession</th>
<th align="right">timepoint</th>
<th align="right">expression_value</th>
<th align="left">replicate</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">0.3968734</td>
<td align="left">Ro18-11-a</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">1.4147711</td>
<td align="left">Ro18-11-b</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">11</td>
<td align="right">0.7423984</td>
<td align="left">Ro18-11-c</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">11.3007002</td>
<td align="left">Ro18-29-a</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">23.2055664</td>
<td align="left">Ro18-29-b</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Ro18</td>
<td align="right">29</td>
<td align="right">22.0307747</td>
<td align="left">Ro18-29-c</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">7</td>
<td align="right">0.4667855</td>
<td align="left">Col0-07-a</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">7</td>
<td align="right">0.0741901</td>
<td align="left">Col0-07-b</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">8</td>
<td align="right">0.0000000</td>
<td align="left">Col0-08-a</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">8</td>
<td align="right">0.0000000</td>
<td align="left">Col0-08-b</td>
</tr>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">9</td>
<td align="right">0.3722542</td>
<td align="left">Col0-09-a</td>
</tr>
<tr class="even">
<td align="left">BRAA02G018970.3C</td>
<td align="left">Col0</td>
<td align="right">9</td>
<td align="right">0.0000000</td>
<td align="left">Col0-09-b</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="registering-the-data">Registering the data<a class="anchor" aria-label="anchor" href="#registering-the-data"></a>
</h3>
<p>To align gene expression time-course between <em>Arabidopsis</em>
Col-0 and <em>B. rapa</em> Ro18, we can use the function
<code><a href="../reference/register.html">register()</a></code>. When using the default
<code>use_optimisation = TRUE</code>, <code>greatR</code> will find the
best stretch and shift parameters through optimisation. For more details
on the other function arguments, go to <code><a href="../reference/register.html">register()</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">registration_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/register.html">register</a></span><span class="op">(</span></span>
<span>  <span class="va">b_rapa_data</span>,</span>
<span>  reference <span class="op">=</span> <span class="st">"Ro18"</span>,</span>
<span>  query <span class="op">=</span> <span class="st">"Col0"</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"z-score"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ── Validating input data ────────────────────────────────────────────────────────</span></span>
<span><span class="co">#&gt; ℹ Will process 10 genes.</span></span>
<span><span class="co">#&gt; ℹ Using estimated standard deviation, as no `exp_sd` was provided.</span></span>
<span><span class="co">#&gt; ℹ Using `scaling_method` = "z-score".</span></span>
<span><span class="co">#&gt;</span></span>
<span><span class="co">#&gt; ── Starting registration with optimisation ──────────────────────────────────────</span></span>
<span><span class="co">#&gt; ℹ Using L-BFGS-B optimisation method.</span></span>
<span><span class="co">#&gt; ℹ Using computed stretches and shifts search space limits.</span></span>
<span><span class="co">#&gt; ℹ Using `overlapping_percent` = 50% as a registration criterion.</span></span>
<span><span class="co">#&gt; ✔ Optimising registration parameters for genes (10/10) [2s]</span></span></code></pre></div>
<p>When dealing with thousands of genes, users may speed up the
registration process by using the argument <code>num_cores</code> to run
the registration using parallel processing.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; 8</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">registration_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/register.html">register</a></span><span class="op">(</span></span>
<span>  <span class="va">b_rapa_data</span>,</span>
<span>  reference <span class="op">=</span> <span class="st">"Ro18"</span>,</span>
<span>  query <span class="op">=</span> <span class="st">"Col0"</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"z-score"</span>,</span>
<span>  num_cores <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="registration-results">Registration results<a class="anchor" aria-label="anchor" href="#registration-results"></a>
</h3>
<p>The function <code><a href="../reference/register.html">register()</a></code> returns a list with S3 class
<code>res_greatR</code> containing three different objects:</p>
<ul>
<li>
<code>data</code> is a data frame containing the expression data and
an additional <code>timepoint_reg</code> column which is a result of
registered time points by applying the registration parameters to the
query data.</li>
<li>
<code>model_comparison</code> is a data frame containing (a) the
optimal stretch and shift for each <code>gene_id</code> and (b) the
difference between Bayesian Information Criterion for the separate model
and for the combined model (<code>BIC_diff</code>) after applying
optimal registration parameters for each gene. If the value of
<code>BIC_diff &lt; 0</code>, then expression dynamics between reference
and query data can be registered (<code>registered = TRUE</code>).
(Default S3 print).</li>
<li>
<code>fun_args</code> is a list of arguments used when calling the
function (<code>reference</code>, <code>query</code>,
<code>scaling_method</code>, …).</li>
</ul>
<p>To check whether a gene is registered or not, we can get the summary
results by accessing the <code>model_comparison</code> table from the
registration result.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">registration_results</span><span class="op">$</span><span class="va">model_comparison</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">gene_id</th>
<th align="right">stretch</th>
<th align="right">shift</th>
<th align="right">BIC_diff</th>
<th align="left">registered</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BRAA02G018970.3C</td>
<td align="right">4.000000</td>
<td align="right">-30.978998</td>
<td align="right">2.830103</td>
<td align="left">FALSE</td>
</tr>
<tr class="even">
<td align="left">BRAA02G043220.3C</td>
<td align="right">2.451533</td>
<td align="right">-10.595789</td>
<td align="right">-3.571541</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">BRAA03G023790.3C</td>
<td align="right">2.250665</td>
<td align="right">-4.358395</td>
<td align="right">-7.849111</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">BRAA03G051930.3C</td>
<td align="right">3.099925</td>
<td align="right">-12.564349</td>
<td align="right">-8.096682</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">BRAA04G005470.3C</td>
<td align="right">3.527640</td>
<td align="right">-20.246481</td>
<td align="right">-7.540003</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">BRAA05G005370.3C</td>
<td align="right">2.275328</td>
<td align="right">-5.030457</td>
<td align="right">-7.726491</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">BRAA06G025360.3C</td>
<td align="right">2.383370</td>
<td align="right">-8.024893</td>
<td align="right">-6.502890</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">BRAA07G030470.3C</td>
<td align="right">4.000000</td>
<td align="right">-27.031933</td>
<td align="right">-5.450813</td>
<td align="left">TRUE</td>
</tr>
<tr class="odd">
<td align="left">BRAA07G034100.3C</td>
<td align="right">4.000000</td>
<td align="right">-27.237508</td>
<td align="right">-3.927756</td>
<td align="left">TRUE</td>
</tr>
<tr class="even">
<td align="left">BRAA09G045310.3C</td>
<td align="right">3.381275</td>
<td align="right">-17.908895</td>
<td align="right">-7.690015</td>
<td align="left">TRUE</td>
</tr>
</tbody>
</table>
<p>From the sample data above, we can see that for nine out of ten
genes, <code>registered = TRUE</code>, meaning that reference and query
data between those nine genes can be aligned or registered. These data
frame outputs can further be summarised and visualised; see the
documentation on the <a href="https://ruthkr.github.io/greatR/articles/process-results.html">processing
registration results</a> article.</p>
</div>
</div>
<div class="section level2">
<h2 id="using-other-inputs">Using other inputs<a class="anchor" aria-label="anchor" href="#using-other-inputs"></a>
</h2>
<div class="section level3">
<h3 id="loading-sample-data-1">Loading sample data<a class="anchor" aria-label="anchor" href="#loading-sample-data-1"></a>
</h3>
<p>As noted in the <a href="https://ruthkr.github.io/greatR/articles/data-requirement.html">input
data requirements</a> article, <code><a href="../reference/register.html">register()</a></code> also accepts a
list of data frames or a list of reference and query vectors as
<code>input</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define expression value vectors</span></span>
<span><span class="va">ref_expressions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.9</span>, <span class="fl">3.1</span>, <span class="fl">7.8</span>, <span class="fl">31.6</span>, <span class="fl">33.7</span>, <span class="fl">31.5</span>, <span class="fl">131.4</span>, <span class="fl">107.5</span>, <span class="fl">116.7</span>, <span class="fl">112.5</span>, <span class="fl">109.7</span>, <span class="fl">57.4</span>, <span class="fl">50.9</span><span class="op">)</span></span>
<span><span class="va">query_expressions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">14</span>, <span class="fl">12.1</span>, <span class="fl">15.9</span>, <span class="fl">47</span>, <span class="fl">30.9</span>, <span class="fl">50.5</span>, <span class="fl">80.1</span>, <span class="fl">67.4</span>, <span class="fl">72.9</span>, <span class="fl">61.7</span><span class="op">)</span></span>
<span></span>
<span><span class="va">list_vector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  reference <span class="op">=</span> <span class="va">ref_expressions</span>,</span>
<span>  query <span class="op">=</span> <span class="va">query_expressions</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="registering-the-data-1">Registering the data<a class="anchor" aria-label="anchor" href="#registering-the-data-1"></a>
</h3>
<p>As previously shown, to register the input list, users can run the
function <code><a href="../reference/register.html">register()</a></code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">registration_results_vectors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/register.html">register</a></span><span class="op">(</span></span>
<span>  <span class="va">list_vector</span>,</span>
<span>  reference <span class="op">=</span> <span class="st">"Ref"</span>,</span>
<span>  query <span class="op">=</span> <span class="st">"Query"</span>,</span>
<span>  scaling_method <span class="op">=</span> <span class="st">"z-score"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ── Validating input data ───────────────────────────────────────────────────────</span></span>
<span><span class="co">#&gt; ℹ Will process 1 gene.</span></span>
<span><span class="co">#&gt; ℹ Using estimated standard deviation, as no `exp_sd` was provided.</span></span>
<span><span class="co">#&gt; ℹ Using `scaling_method` = "z-score".</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Starting registration with optimisation ─────────────────────────────────────</span></span>
<span><span class="co">#&gt; ℹ Using L-BFGS-B optimisation method.</span></span>
<span><span class="co">#&gt; ℹ Using computed stretches and shifts search space limits.</span></span>
<span><span class="co">#&gt; ℹ Using `overlapping_percent` = 50% as a registration criterion.</span></span>
<span><span class="co">#&gt; ✔ Optimising registration parameters for genes (1/1) [170ms]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="registration-results-1">Registration results<a class="anchor" aria-label="anchor" href="#registration-results-1"></a>
</h3>
<p>The registration result object will have the same structure as when
using a data frame as an input. Since no ID is explicitly defined in the
input vector list, a unique <code>gene_id</code> will be automatically
generated for the reference and query pair.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">registration_results_vectors</span><span class="op">$</span><span class="va">model_comparison</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">gene_id</th>
<th align="right">stretch</th>
<th align="right">shift</th>
<th align="right">BIC_diff</th>
<th align="left">registered</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">5472143a</td>
<td align="right">1.125762</td>
<td align="right">0.074995</td>
<td align="right">-6.115123</td>
<td align="left">TRUE</td>
</tr></tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://ruthkr.com" class="external-link">Ruth Kristianingsih</a>, Alex Calderwood.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
